Backends/CLX/clipboard.lisp

```diff
@@ -0,0 +1,296 @@
+(in-package :clim-clx)
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; CLX implementation of clipboard management
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+;;; Port mixin
+
+(defclass clx-clipboard-stored-object ()
+  ((content :initarg :content
+            :reader clipboard-stored-object-content)
+   (type    :initarg :type
+            :reader clipboard-stored-object-type)
+   (owner   :initarg :owner
+            :reader clipboard-stored-object-owner)))
+
+(defclass clx-clipboard-port-mixin ()
+  ((selection-content :initform nil
+                      :accessor selection-content)
+   (clipboard-content :initform nil
+                      :accessor clipboard-content)
+   (outstanding-request-pane :initform nil
+                             :accessor clipboard-outstanding-request-pane)
+   (outstanding-request-type  :initform nil
+                              :accessor clipboard-outstanding-request-type)
+   (outstanding-request-selection :initform nil
+                                  :accessor clipboard-outstanding-request-selection)))
+
+(defun find-stored-object (port selection)
+  (ecase selection
+    (:primary (selection-content port))
+    (:clipboard (clipboard-content port))))
+
+(defun set-stored-object (port selection value)
+  (ecase selection
+    (:primary (setf (selection-content port) value))
+    (:clipboard (setf (clipboard-content port) value))))
+
+;;; Event classes
+
+(defclass clx-selection-event (window-event)
+  ())
```

Comment:
get rid of all this event hackery and put calls to process-selection-* directly in event-handler in input.lisp.

Reply:
The event hackery originates from the old implementation. I decided to keep it since I wanted to make sure the clipboard content accessed from the main application thread, and not the event thread (in order to avoid race conditions).

I now note that even the `HANDLE-EVENT` is processed in the main CLX event loop rather than the application thread. Clearly this was a problem even with the old implementation.

This means that moving everything to `input.lisp` is definitely possible, and will not cause any more problems than we're already facing. However, the threading problem exists, and I don't know what should be done to fix it.

I'm pretty sure there are other places where objects are accessed from both the event thread and the application thread without locking (I'd presume anything that implements `HANDLE-EVENT`).

Reply:
handle-event is always issues from the application event loop. event queue is a thread-safe mailbox, so queue and handle are not prone to races of any kind. that doesn't change the fact that handling this communication with X is local to the port (and only send-notify is passed to the rest of the system). For setting clipboard from the application thread – this should be made secure thread-wise.

I still maintain that this handle-event hackery should be removed.

Reply:
After actually removing the code that implemented the hackery, I agree with you. It's much cleaner this way, and doesn't introduce any new concurrency issues that didn't already exist.

I've pushed a fix.